examples/device-allocator.cpp

```diff
@@ -4,29 +4,81 @@
+#include "camp/camp.hpp"
 #include "umpire/ResourceManager.hpp"
+#include "umpire/device_allocator_helper.hpp"
 
-__global__ void my_kernel(umpire::DeviceAllocator alloc, double** data_ptr)
+#if defined(UMPIRE_ENABLE_CUDA)
+using resource_type = camp::resources::Cuda;
+#elif defined(UMPIRE_ENABLE_HIP)
+using resource_type = camp::resources::Hip;
+#endif
+
+/*
+ * Very simple kernel that uses only the first thread to "get" the
+ * existing DeviceAllocator and to allocate an array of ints.
+ * Making sure that the data_ptr is pointing to the device allocated array,
+ * it sets one element to a value that will be checked later.
+ */
+__global__ void my_kernel(double** data_ptr)
 {
   if (threadIdx.x == 0) {
+    // _sphinx_tag_get_dev_allocator_id_start
+    umpire::DeviceAllocator alloc = umpire::getDeviceAllocator(0);
+    // _sphinx_tag_get_dev_allocator_id_end
     double* data = static_cast<double*>(alloc.allocate(10 * sizeof(double)));
     *data_ptr = data;
-
     data[7] = 1024;
   }
 }
 
+__global__ void my_other_kernel(double** data_ptr)
+{
+  if (threadIdx.x == 0) {
+    // _sphinx_tag_get_dev_allocator_name_start
+    umpire::DeviceAllocator alloc = umpire::getDeviceAllocator("my_other_device_alloc");
+    // _sphinx_tag_get_dev_allocator_name_end
+    double* data = static_cast<double*>(alloc.allocate(1 * sizeof(double)));
+    *data_ptr = data;
+    data[0] = 42;
+  }
+}
+
 int main(int argc, char const* argv[])
 {
   auto& rm = umpire::ResourceManager::getInstance();
+  auto resource = camp::resources::Resource{resource_type{}};
+
+  // Create all of my allocators
   auto allocator = rm.getAllocator("UM");
-  auto device_allocator = umpire::DeviceAllocator(allocator, 1024);
+  // _sphinx_tag_make_dev_allocator_start
+  auto device_allocator = umpire::makeDeviceAllocator(allocator, 1024, "my_device_alloc");
+  auto device_allocator2 = umpire::makeDeviceAllocator(allocator, 512, "my_other_device_alloc");
+  // _sphinx_tag_make_dev_allocator_end
+
+  // Checking that now a DeviceAllocator exists
+  int id_found = umpire::findDeviceAllocatorID("my_device_alloc");
+  if (umpire::deviceAllocatorExists(id_found)) {
+    std::cout << "I found a DeviceAllocator with ID: " << id_found << std::endl;
+  }
 
   double** ptr_to_data = static_cast<double**>(allocator.allocate(sizeof(double*)));
 
-  my_kernel<<<1, 16>>>(device_allocator, ptr_to_data);
-  cudaDeviceSynchronize();
+  // Make sure that device and host side DeviceAllocator pointers are synched
+  // _sphinx_tag_macro_start
+  UMPIRE_SET_UP_DEVICE_ALLOCATORS();
+  // _sphinx_tag_macro_end
+
+  my_kernel<<<1, 16>>>(ptr_to_data);
+  resource.get_event().wait();
+  std::cout << "After first kernel, found value: " << (*ptr_to_data)[7] << std::endl;
+
+  my_other_kernel<<<1, 16>>>(ptr_to_data);
+  resource.get_event().wait();
+  std::cout << "After second kernel, found value: " << (*ptr_to_data)[0] << std::endl;
+
+  // Tear down and deallocate memory.
+  umpire::destroyDeviceAllocator();
```

Comment:
This function should be called by the ResourceManager destructor, not manually by the user.

Reply:
There was a discussion at one point that we wanted to eventually move away from the ResourceManager. This will make it go from being completely independent of the RM to not (so it will be created by the user manually not using the RM, but destroyed automatically with the RM.) Is that ok?

Reply:
Good point. If we remove the resource manager from the user interface we will still have some kind of "shutdown" mechanism to do stuff like this, so I think moving it for now is a good idea so we don't need the user to call this function.

Reply:
Mkay!